lib/puppet/pops/lookup/invocation.rb

```diff
@@ -67,13 +75,14 @@ def lookup(key, module_name = nil)
     @top_key = key
     @module_name = module_name.nil? ? key.module_name : module_name
     save_current = self.class.current
-    if save_current.equal?(self)
+    if save_current && save_current.value.equal?(self)
       yield
     else
       begin
-        self.class.instance_variable_set(:@current, self)
+        self.class.instance_variable_set(:@current, Puppet::ThreadLocal.new(self))
```

Comment:
IIRC, this will set the global value for current. Setting the threadlocal value would be done by calling `self.class.current.value = self`, I think.

Reply:
Thank